interfaces/acados_template/acados_template/mpc_utils.py

```diff
@@ -79,41 +81,250 @@ def create_model_with_cost_state(ocp: AcadosOcp) -> Tuple[AcadosModel, np.ndarra
         for ig in ocp.constraints.idxsg:
             lower_violation = ca.fmax(ocp.constraints.lg[ig] - g[ig], 0)
             upper_violation = ca.fmax(g[ig] - ocp.constraints.ug[ig], 0)
-            cost_dot += ocp.cost.zl[i_slack] * lower_violation + ocp.cost.Zl[i_slack] * lower_violation ** 2
-            cost_dot += ocp.cost.zu[i_slack] * upper_violation + ocp.cost.Zu[i_slack] * upper_violation ** 2
+            cost_dot += ocp.cost.zl[i_slack] * lower_violation + \
+                ocp.cost.Zl[i_slack] * lower_violation ** 2
+            cost_dot += ocp.cost.zu[i_slack] * upper_violation + \
+                ocp.cost.Zu[i_slack] * upper_violation ** 2
             i_slack += 1
 
     for ih in ocp.constraints.idxsh:
-        lower_violation = ca.fmax(ocp.constraints.lh[ih] - ocp.model.con_h_expr[ih], 0)
-        upper_violation = ca.fmax(ocp.model.con_h_expr[ih] - ocp.constraints.uh[ih], 0)
-        cost_dot += ocp.cost.zl[i_slack] * lower_violation + ocp.cost.Zl[i_slack] * lower_violation ** 2
-        cost_dot += ocp.cost.zu[i_slack] * upper_violation + ocp.cost.Zu[i_slack] * upper_violation ** 2
+        lower_violation = ca.fmax(
+            ocp.constraints.lh[ih] - ocp.model.con_h_expr[ih], 0)
+        upper_violation = ca.fmax(
+            ocp.model.con_h_expr[ih] - ocp.constraints.uh[ih], 0)
+        cost_dot += ocp.cost.zl[i_slack] * lower_violation + \
+            ocp.cost.Zl[i_slack] * lower_violation ** 2
+        cost_dot += ocp.cost.zu[i_slack] * upper_violation + \
+            ocp.cost.Zu[i_slack] * upper_violation ** 2
         i_slack += 1
 
     if not is_empty(ocp.constraints.idxsphi):
-        raise NotImplementedError(f"Not implemented for nontrivial ocp.constraints.idxsphi = {ocp.constraints.idxsphi}")
+        raise NotImplementedError(
+            f"Not implemented for nontrivial ocp.constraints.idxsphi = {ocp.constraints.idxsphi}")
 
     model.x = ca.vertcat(model.x, cost_state)
     model.xdot = ca.vertcat(model.xdot, cost_state_dot)
     model.f_expl_expr = ca.vertcat(model.f_expl_expr, cost_dot)
     model.f_impl_expr = ca.vertcat(model.f_impl_expr, cost_state_dot-cost_dot)
 
     return model, ocp.parameter_values
+
+
+def detect_constraint_structure(model: AcadosModel, constraints: AcadosOcpConstraints, stage_type: str):
+    """
+    - stage_type: allowed values "initial", "path", "terminal"
+    """
+    x = model.x
+    u = model.u
+    z = model.z
+    nx = x.shape[0]
+    nu = u.shape[0]
+
+    if model.p_global is None:
+        p_global = []  # to have same structure of model.p
+
+    if isinstance(x, ca.SX):
+        CASADI_VAR = ca.SX
+    elif isinstance(x, ca.MX):
+        CASADI_VAR = ca.MX
+    else:
+        raise TypeError(
+            "Optimization variables 'x' must be of type 'casadi SX' or 'casadi MX'")
+
+    if stage_type == 'initial':
+        expr_constr = model.con_h_expr_0
+        LB = constraints.lh_0
```

Comment:
I would avoid all upper case, in Python this should be used for global variables.